caas/kubernetes/provider/k8s.go

```diff
@@ -769,6 +855,84 @@ func (k *kubernetesClient) DeleteOperator(appName string) (err error) {
 	return errors.Trace(k.deleteDeployment(operatorName))
 }
 
+func getSvcPort(svc *core.Service, portName string) (int32, error) {
+	choosePort := func(svcType core.ServiceType, p core.ServicePort) int32 {
+		switch svcType {
+		case core.ServiceTypeNodePort:
+			return p.NodePort
+		default:
+			return p.Port
+		}
+	}
+
+	var port int32
+	for _, p := range svc.Spec.Ports {
+		if p.Name == portName {
+			return choosePort(svc.Spec.Type, p), nil
+
+		}
+	}
+	return port, errors.NotFoundf("port %q in service %q", portName, svc.Name)
+}
+
+func getSvcAddresses(svc *core.Service) []network.Address {
+	var netAddrs []network.Address
+
+	appendAddrs := func(scope network.Scope, addrs ...string) {
+		for _, v := range addrs {
+			if v != "" {
+				netAddrs = append(netAddrs, network.Address{
+					Value: v,
+					Type:  network.DeriveAddressType(v),
+					Scope: scope,
+				})
+			}
+		}
+	}
+	scope := network.ScopePublic
+	clusterIP := svc.Spec.ClusterIP
+	t := svc.Spec.Type
+	switch t {
+	case core.ServiceTypeClusterIP:
+		appendAddrs(network.ScopeCloudLocal, clusterIP)
+	case core.ServiceTypeExternalName:
+		appendAddrs(network.ScopeCloudLocal, svc.Spec.ExternalName)
+	case core.ServiceTypeNodePort:
+		appendAddrs(scope, svc.Spec.ExternalIPs...)
+	case core.ServiceTypeLoadBalancer:
+		appendAddrs(scope, svc.Spec.LoadBalancerIP)
+	}
+	if len(netAddrs) == 0 && clusterIP != "" {
+		// fallback to ClusterIP, usually it's not empty.
+		logger.Debugf("fallback to clusterIP %q, desired IP was empty for %q type service %q", clusterIP, t, svc.Name)
+		appendAddrs(network.ScopeCloudLocal, clusterIP)
+	}
+	return netAddrs
+}
+
+// GetServicePublicAddresses returns the addresses of the service.
+func (k *kubernetesClient) GetServicePublicAddresses(svcName string) ([]network.Address, error) {
```

Comment:
We should just add these attribute to the Service object and call the existing GetService()

Reply:
ah, right.